src/useAnswersState.tsx

```diff
@@ -1,15 +1,75 @@
-import { useContext, useRef, useState } from 'react';
+import React, { useCallback, useContext, useEffect, useRef, useState } from 'react';
 import { State } from '@yext/answers-headless';
 import { AnswersHeadlessContext } from './AnswersHeadlessContext';
 import useLayoutEffect from 'use-isomorphic-layout-effect';
+import { useSyncExternalStoreWithSelector } from 'use-sync-external-store/shim/with-selector';
 
 export type StateSelector<T> = (s: State) => T;
 
+const canUseDOM = !!(typeof window !== 'undefined'
+  && typeof window.document !== 'undefined'
+  && typeof window.document.createElement !== 'undefined');
+const isServerEnvironment = !canUseDOM;
+
+/**
+ * use-sync-external-store/shim does not support getServerSnapshot for pre-18 React versions.
+ * It requires user of the shim to handle subscription logic and return the correct snapshot value.
+ * As such, server side rendering from pre-18 React version will use a version of useAnswersState that
+ * manually handle memoization of state selector and register subscription without useSyncExternalStore call.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const useAnswersState = (React as any).useSyncExternalStore === undefined && isServerEnvironment
+  ? useAnswersStateWithManualStoreSync : useAnswersStateWithReactStoreSync;
+
+
+/**
+ * Returns the Answers State returned by the map function.
+ * UseSyncExternalStoreWithSelector is used to subscribe to external store and trigger rerender
+ * whenever an action is dispatched and caused changes to the state of the selector function.
+ */
+function useAnswersStateWithReactStoreSync<T>(stateSelector: StateSelector<T>): T {
```

Comment:
would it make sense to separate these two functions into separate files?

Reply:
removed `useAnswersStateWithManualSyncStore` so this is no longer needed